Matrix generator's getParams should check error return to avoid panic

When the Matrix generator returns an error (for example, open of the child generators failed to successfully complete), the error is ignored and an array is indexed out-of-bounds. See full error message below.


As reported on Slack:
https://cloud-native.slack.com/archives/C01U45M2SVB/p1627898541004300

Log:
```
time="2021-08-02T09:51:15Z" level=error msg="error generating params" error="Error during fetching commitSHA: dial tcp 140.82.112.3:22: connect: connection timed out" generator="&{0xc000b26960}"
panic: runtime error: index out of range [0] with length 0
goroutine 300 [running]:
github.com/argoproj-labs/applicationset/pkg/generators.(*MatrixGenerator).getParams(0xc00000e718, 0x0, 0x0, 0xc000d4d1e0, 0x0, 0x0, 0xc0005d9680, 0xc001089b88, 0x4, 0xc000b8c780, ...)
        /workspace/pkg/generators/matrix.go:79 +0x15c
github.com/argoproj-labs/applicationset/pkg/generators.(*MatrixGenerator).GenerateParams(0xc00000e718, 0xc002816a20, 0xc0005d9680, 0xc000624100, 0x7b, 0x0, 0x0, 0xc002816930)
        /workspace/pkg/generators/matrix.go:41 +0xbd
github.com/argoproj-labs/applicationset/pkg/generators.Transform(0x0, 0x0, 0x0, 0xc002620f20, 0x0, 0x0, 0xc000b19170, 0xc000624100, 0x7b, 0x0, ...)
        /workspace/pkg/generators/generator_spec_processor.go:52 +0x50b
github.com/argoproj-labs/applicationset/pkg/controllers.(*ApplicationSetReconciler).generateApplications(0xc000b25cb0, 0x21c5302, 0xe, 0xc0031d95a8, 0x14, 0xc000dc3da0, 0x19, 0x0, 0x0, 0xc001089ab0, ...)
        /workspace/pkg/controllers/applicationset_controller.go:220 +0x205
github.com/argoproj-labs/applicationset/pkg/controllers.(*ApplicationSetReconciler).Reconcile(0xc000b25cb0, 0x2b23ed8, 0xc002816750, 0xc000c52480, 0x4, 0xc000e8d760, 0x19, 0xc002816750, 0xc00003a000, 0x258b3e0, ...)
        /workspace/pkg/controllers/applicationset_controller.go:93 +0x298
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000b50820, 0x2b23e30, 0xc000cb6000, 0x24f8560, 0xc001883480)
        /workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:298 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000b50820, 0x2b23e30, 0xc000cb6000, 0x0)
        /workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2(0xc000dc4010, 0xc000b50820, 0x2b23e30, 0xc000cb6000)
        /workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:214 +0x6b
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /workspace/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:210 +0x425
```

--- a/pkg/generators/matrix.go
+++ b/pkg/generators/matrix.go
@@ -2,6 +2,7 @@ package generators
 
 import (
 	"errors"
+	"fmt"
 	"time"
 
 	argoprojiov1alpha1 "github.com/argoproj-labs/applicationset/api/v1alpha1"
@@ -62,7 +63,7 @@ func (m *MatrixGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.App
 
 func (m *MatrixGenerator) getParams(appSetBaseGenerator argoprojiov1alpha1.ApplicationSetBaseGenerator, appSet *argoprojiov1alpha1.ApplicationSet) ([]map[string]string, error) {
 
-	t, _ := Transform(
+	t, err := Transform(
 		argoprojiov1alpha1.ApplicationSetGenerator{
 			List:     appSetBaseGenerator.List,
 			Clusters: appSetBaseGenerator.Clusters,
@@ -71,6 +72,14 @@ func (m *MatrixGenerator) getParams(appSetBaseGenerator argoprojiov1alpha1.Appli
 		m.supportedGenerators,
 		argoprojiov1alpha1.ApplicationSetTemplate{},
 		appSet)
+
+	if err != nil {
+		return nil, fmt.Errorf("child generator returned an error on parameter generation: %v", err)
+	}
+
+	if len(t) == 0 {
+		return nil, fmt.Errorf("child generator generated no parameters")
+	}
 
 	if len(t) > 1 {
 		return nil, MoreThenOneInnerGenerators

